src/lib/coda_base/user_command_memo.ml

```diff
@@ -1,21 +1,168 @@
+
+let length memo = Char.to_int memo.[length_index]
+
+let is_digest memo = Char.equal (tag memo) digest_tag
+
+let is_valid memo =
+  Int.equal (String.length memo) (digest_length + 2)
+  &&
+  let memo_len = length memo in
+  if is_digest memo then Int.equal memo_len digest_length
+  else
+    Char.equal (tag memo) bytes_tag
+    && Int.(memo_len <= digest_length)
+    &&
+    let padded =
+      String.sub memo ~pos:(memo_len + 2) ~len:(digest_length - memo_len)
+    in
+    String.for_all padded ~f:(Char.equal '\x00')
+
+let create_by_digesting_string_exn s =
+  if Int.(String.length s > max_digestible_string_length) then
+    raise Too_long_digestible_string ;
+  let digest = (Random_oracle.digest_string s :> t) in
+  String.init (digest_length + 2) ~f:(fun ndx ->
+      if Int.equal ndx tag_index then digest_tag
+      else if Int.equal ndx length_index then digest_length_byte
+      else digest.[ndx - 2] )
+
+let create_by_digesting_string (s : string) =
+  try Ok (create_by_digesting_string_exn s)
+  with Too_long_digestible_string ->
+    Or_error.error_string "create_by_digesting_string: string too long"
+
+module type Memoable = sig
+  type t
+
+  val length : t -> int
+
+  val get : t -> int -> char
+end
+
+let create_from_value_exn (type t) (module M : Memoable with type t = t)
+    (value : t) =
+  let len = M.length value in
+  if not Int.(len <= digest_length) then raise Too_long_user_memo_length ;
+  String.init (digest_length + 2) ~f:(fun ndx ->
+      if Int.equal ndx tag_index then bytes_tag
+      else if Int.equal ndx length_index then Char.of_int_exn len
+      else if Int.(ndx <= len + 2) then M.get value (ndx - 2)
+      else '\x00' )
+
+let create_from_bytes_exn bytes = create_from_value_exn (module Bytes) bytes
+
+let create_from_bytes bytes =
+  try Ok (create_from_bytes_exn bytes)
+  with Too_long_user_memo_length ->
+    Or_error.error_string
+      (sprintf "create_from_bytes: length exceeds %d" digest_length)
+
+let create_from_string_exn s = create_from_value_exn (module String) s
+
+let create_from_string s =
+  try Ok (create_from_string_exn s)
+  with Too_long_user_memo_length ->
+    Or_error.error_string
+      (sprintf "create_from_string: length exceeds %d" digest_length)
+
+let dummy = (create_by_digesting_string_exn "" :> t)
+
+module Boolean = Tick0.Boolean
+module Typ = Tick0.Typ
+
+(* the code below is much the same as in Random_oracle.Digest; tag and length bytes 
+   make it a little different
+ *)
+
+module Checked = struct
+  type unchecked = t
+
+  type t = Boolean.var array
+
+  let to_triples t =
+    Fold_lib.Fold.(to_list (group3 ~default:Boolean.false_ (of_array t)))
+
+  let constant unchecked =
+    assert (Int.(String.length (unchecked :> string) = digest_length + 2)) ;
+    Array.map
+      (Blake2.string_to_bits (unchecked :> string))
+      ~f:Boolean.var_of_value
+end
+
+let length_in_bits = 8 * digest_length
+
+let length_in_triples = (length_in_bits + 2) / 3
```

Comment:
I see this `+2` is also in the `random_oracle_base.ml`, I guess to make sure the length in triples gets rounded up and not down by the integer division?

Reply:
I believe so.